### PR TITLE
Upgrade to piecrust `0.5`

### LINKED ISSUE #883: Update to `piecrust` version `0.5`
#### Summary
This is necessary to allow the registration of contract `owners`, which in turn is necessary for implementing the [fee market](#880).

#### Possible solution design or implementation
```toml
piecrust = "0.5"
piecrust-uplink = "0.5"
```

### PR DESCRIPTION
**Pull request by ureeves**
Resolves #883

### CODE REVIEW

File: rusk/src/lib/lib.rs [hunk 1]
```diff
@@ -396,7 +402,7 @@ impl Rusk {
 
     fn query<A, R>(
         &self,
-        module_id: ModuleId,
+        module_id: ContractId,
```
Comment (HDauven) [inline rusk/src/lib/lib.rs:405]: Should we call the `module_id` herr `contract_id` too or is there a reason to keep the module nomenclature in this context?

**Review by HDauven — APPROVED** [review]
LGTM, please see comments though.

File: rusk/src/lib/lib.rs (hunk 1 above)
Comment (ureeves) [reply rusk/src/lib/lib.rs:405]: This is clearly a stray. Will change.

File: contracts/stake/src/state.rs [hunk 2]
```diff
@@ -183,20 +181,19 @@ impl StakeState {
...
+                note: unstake.note,
+                proof: unstake.proof,
+            },
+        )
+        .expect("Withdrawing note from contract should be successful");
     }
 
-    pub fn withdraw(self: &mut State<Self>, withdraw: Withdraw) {
+    pub fn withdraw(&mut self, withdraw: Withdraw) {
         if rusk_abi::caller() != rusk_abi::transfer_module() {
```
Comment (herr-seppia) [inline contracts/stake/src/state.rs]: Out of scope:
Should this check verify also that the caller is the "root" caller?

File: contracts/stake/tests/stake.rs [hunk 3]
```diff
@@ -737,11 +748,15 @@ fn stake_withdraw_unstake() {
 
     // set different block height so that the new notes are easily located and
     // filtered
-    rusk_abi::set_block_height(session, 3);
+    // sets the block height for all subsequent operations to 1
+    let base = session.commit().expect("Committing should succeed");
+    let mut session = rusk_abi::new_session(vm, base, 3)
+        .expect("Instantiating new session should succeed");
+    session.set_point_limit(POINT_LIMIT);
```
Comment (herr-seppia) [inline contracts/stake/tests/stake.rs:750]: Is this required? You're setting point_limit 2 lines after too

File: contracts/transfer/tests/transfer.rs [hunk 4]
```diff
@@ -62,40 +62,46 @@ fn instantiate<Rng: RngCore + CryptoRng>(
...
-        .transact(rusk_abi::transfer_module(), "update_root", &())
+        .call(TRANSFER_CONTRACT, "update_root", &())
         .expect("Updating the root should succeed");
 
     // sets the block height for all subsequent operations to 1
-    rusk_abi::set_block_height(&mut session, 1);
+    let base = session.commit().expect("Committing should succeed");
+    let mut session = rusk_abi::new_session(vm, base, 1)
+        .expect("Instantiating new session should succeed");
+    session.set_point_limit(POINT_LIMIT);
```
Comment (herr-seppia) [inline contracts/transfer/tests/transfer.rs:104]: This is not clear to me.
Previous version didn't require to commit before return. Is something changed internally with `0.5`?

File: contracts/alice/src/state.rs [hunk 5]
```diff
@@ -16,21 +16,18 @@ impl Alice {
...
+        let _: bool = rusk_abi::call(TRANSFER_CONTRACT, "wfco", &wfco)
             .expect("Obfuscated withdrawal transaction should succeed");
     }
 
-    pub fn withdraw_to_contract(self: &mut State<Self>, wfctc: Wfctc) {
-        let _: bool = self
-            .transact(rusk_abi::transfer_module(), "wfctc", &wfctc)
+    pub fn withdraw_to_contract(&mut self, wfctc: Wfctc) {
+        let _: bool = rusk_abi::call(TRANSFER_CONTRACT, "wfctc", &wfctc)
             .expect("Obfuscated withdrawal transaction should succeed");
```
Comment (herr-seppia) [inline contracts/alice/src/state.rs:31]: Change the panic message to not be equal to `withdraw_obfuscated` one

File: rusk-recovery/src/state/snapshot/governance.rs [hunk 6]
```diff
@@ -6,32 +6,38 @@
...
+use dusk_pki::{PublicKey, PublicSpendKey};
+use rusk_abi::ContractId;
 use serde_derive::{Deserialize, Serialize};
 
 use super::wrapper::Wrapper;
 
 #[derive(Serialize, Deserialize, PartialEq, Eq)]
 pub struct Governance {
+    pub(crate) contract_owner:
+        Wrapper<PublicSpendKey, { PublicSpendKey::SIZE }>,
```
Comment (herr-seppia) [inline rusk-recovery/src/state/snapshot/governance.rs:18]: Can you update the `example_gov.toml` to support this mandatory field?

**Review by herr-seppia — CHANGES_REQUESTED** [review]
LGTM overall

Still some nits to address before approval

File: contracts/stake/tests/stake.rs (hunk 3 above)
Comment (ureeves) [reply contracts/stake/tests/stake.rs:750]: Clearly a copy paste issue. Will fix

File: rusk-recovery/src/state/snapshot/governance.rs (hunk 6 above)
Comment (herr-seppia) [reply rusk-recovery/src/state/snapshot/governance.rs:18]: To be on the same page
You didn't make it always mandatory, it's mandatory only if you add any governance entry

File: contracts/transfer/tests/transfer.rs (hunk 4 above)
Comment (ureeves) [reply contracts/transfer/tests/transfer.rs:104]: Yes. We can't set session data - such as block height - after the session has started.

File: contracts/alice/src/state.rs (hunk 5 above)
Comment (ureeves) [reply contracts/alice/src/state.rs:31]: :+1:

**Review by herr-seppia — APPROVED** [review]
LGTM

**Pull request merged**